### PR TITLE
ColorPicker Undo

### PR DESCRIPTION
**Pull request by Gytaco**
Here is the correct branch with the Colorpicker undo functionality

### CODE REVIEW

**Comment by Gytaco** [comment]
Hey Mike here is the colorpicker undo. Please note the reason the colors are setup as RGBA rather then ARGB is due to the setup of the Color.cs class in Dynamo in CoreNodes. Please check the code and you will see the reason why I have to set things to the RGBA for correct coloring and functionality. Also note with the undo if we need the first undo to do 2 undo's we need to add some additional code to manage the additional change. Let me know if your happy with the current setup and I can add the additional functionality.

**Comment by gregmarr** [comment]
The problem isn't that the format is ARGB, the problem is that the Format argument string says to expect `R, G, B, A` in the remaining arguments:
```
Alpha = {3}, Red = {0}, Green = {1}, Blue = {2}
```
but the arguments are in the order `A, R, G, B`
```
DsColor.Alpha, DsColor.Red, DsColor.Green, DsColor.Blue
```
The string should be
```
Alpha = {0}, Red = {1}, Green = {2}, Blue = {3}
```

**Comment by Gytaco** [comment]
Sorry Greg, so due to the format when i save to xml and have to reopen on line 85 I have to pull the data back in as RGBA, so the ToString was to standardise that same output since the node actually uses the same DsColor class. However since they both use the same class the ToString is actually not used at all and I have since removed it so thanks for pointing that out!

File: src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/ColorPalette.cs [hunk 1]
```diff
@@ -16,33 +16,32 @@ public class ColorPaletteNodeViewCustomization : NotificationObject, INodeViewCu
...
         ///     Customize View.
         /// </summary>
         /// <param name="model"></param>
         /// <param name="nodeView"></param>
         public void CustomizeView(ColorPalette model, NodeView nodeView)
         {
+            viewNode = nodeView;
             colorPaletteNode = model;
-            mcolor = Color.FromArgb(model.dsColor.Alpha, model.dsColor.Red, model.dsColor.Green, model.dsColor.Blue);
+            model.PropertyChanged += Model_PropertyChanged;
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/ColorPalette.cs]: I think you should unsubscribe this handler in the dispose method below.

File: src/Libraries/CoreNodeModels/Input/ColorPalette.cs [hunk 2]
```diff
@@ -25,18 +29,46 @@ namespace CoreNodeModels.Input
...
+        {
+            set
+            {
+                OnPropertyChanged();
             }
         }
+
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Input/ColorPalette.cs]: while this is nice, lets try to match all other onPropertyChanged handlers in the repo and just make this the property name string, drop the attribute, and just use the string and raise this directly with the string as an explicit parameter.

File: src/Libraries/CoreNodeModels/Input/ColorPalette.cs [hunk 3]
```diff
@@ -1,9 +1,12 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.ComponentModel;
 using System.Linq;
 using System.Text.RegularExpressions;
 using System.Xml;
 
+using System.Runtime.CompilerServices;
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Input/ColorPalette.cs]: shouldn't need this if you remove the callerMember attribute

File: src/Libraries/CoreNodeModels/Input/ColorPalette.cs [hunk 4]
```diff
@@ -25,18 +29,46 @@ namespace CoreNodeModels.Input
     [OutPortNames("Color")]
     [OutPortTypes("Color")]
     [OutPortDescriptions("Selected Color.")]
-    public class ColorPalette : NodeModel
+    public class ColorPalette : NodeModel, INotifyPropertyChanged
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Input/ColorPalette.cs]: I don't think you need this interface as NodeModel should already implement INotifyPropertyChanged

File: src/Libraries/CoreNodeModels/Input/ColorPalette.cs [hunk 5]
```diff
@@ -25,18 +29,46 @@ namespace CoreNodeModels.Input
...
+                }
                 dscolor = value;
+                if (!Isundo)
+                {
+                    //If undo recorder is not being used updates the undorecorder stack with new value
+                    Isundo = false;
+                    Update = dscolor;
+                }
                 OnNodeModified();
+                OnPropertyChanged();
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Input/ColorPalette.cs]: call this with the property name explicitly

File: src/Libraries/CoreNodeModels/Input/ColorPalette.cs [hunk 6]
```diff
@@ -25,18 +29,46 @@ namespace CoreNodeModels.Input
...
-        public DSColor dsColor
+        private DSColor dscolor = DSColor.ByARGB(255, 0, 0, 0);
+        private bool Isundo = false;
+        public DSColor DsColor
         {
             get { return dscolor; }
             set
-            {                              
+            {
+                if (dscolor.Equals(value))
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Input/ColorPalette.cs]: this logic requires some comments

File: src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/ColorPalette.cs [hunk 7]
```diff
@@ -16,33 +16,32 @@ public class ColorPaletteNodeViewCustomization : NotificationObject, INodeViewCu
...
+            var undoRecorder = viewNode.ViewModel.WorkspaceViewModel.Model.UndoRecorder;
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/ColorPalette.cs]: any reason you are recording the state of the color picker immediately? When the user first selects a color - that should be the time you record the **old** color. Maybe this has to do with the two first undo bug you are seeing.

**Comment by mjkkirschner** [comment]
@Gytaco thanks for this work - I have a few comments for you to check out - also - if possible it would be great to add an undo test for this node.

very similar to your other test, except invoke undo and asset the value is correct.

**Comment by mjkkirschner** [comment]
@Gytaco you can see an example here:
https://github.com/DynamoDS/Dynamo/blob/master/test/DynamoCoreTests/PresetsTests.cs#L192

**Comment by Gytaco** [comment]
Hey @mjkkirschner here is the updated version. Are you happy with the type of hack?? I'm not but be interested to here if you have better ideas. Otherwise we can expand on this idea for the undo? Once I've got this sorted I'll write the unit tests.

**Pull request merged**